streams: Remove get_sub calls and use stream id instead of stream name.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This PR has 3 commits - 
1. Change is_user_subscribed to use stream id instead of stream name.
2. Use narrow_state.stream_sub instead of narrow_state.stream for subscribing to a stream from stream narrow.
3. Store list of stream ids instead of list of stream names which will then be used sorting of streams.

 <!-- How have you tested? -->


<!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented:
These tests seem to not pass arguments improperly; can you look into them?
```
(zulip-py3-venv) tabbott@coset:~/zulip$ git grep needs_subscribe_warn
frontend_tests/node_tests/compose.js:run_test('needs_subscribe_warning', () => {
frontend_tests/node_tests/compose.js:    assert.equal(compose.needs_subscribe_warning(), false);
frontend_tests/node_tests/compose.js:    assert.equal(compose.needs_subscribe_warning(), false);
frontend_tests/node_tests/compose.js:    assert.equal(compose.needs_subscribe_warning(), false);
frontend_tests/node_tests/compose.js:    assert.equal(compose.needs_subscribe_warning(), true);
```

timabbott commented:
I merged the first 2 commits but not the 3rd as it's a bit more complex.

sahil839 commented:
Fixed the tests to pass stream-id and user-id as arguments to `needs_subscribe_warning`.

I have also removed the stubs of `people.get_by_user_id` and `stream_data.is_user_subscribed`, as now we are passing the correct arguments, it makes sense to test this function with original code.

sahil839 commented:
Made a new PR for test changes #15816 and removed that commit from here, as this is tagged with post-release.

timabbott commented:
@sahil839 can you split this into a few more commits?  It seems to have a couple fundamentally separate ideas mixed together (E.g. the `stream_data` and `stream_list` changes seem independent of the rest?)

sahil839 commented:
@timabbott I have made a separate commit for changes in `stream_data.js`. The changes in `stream_list.js` are related to `stream_sort.js` as that is the place from where we send stream ids instead of names to `stream_sort.sort_groups` so I have kept these changes in the same commit.

timabbott commented:
I think the `stream_list.js` is at least logically independent, but it's not a big deal.  Merged, thanks @sahil839!

timabbott commented:
How much more is less for purging unnecessary use of `get_sub`?

sahil839 commented:
I am certainly sure that we can remove 2 or 3 more of them, but have to do a careful audit of the others as most of them are used where we get stream name from compose_state.stream_name, will inform again by tomorrow.

timabbott commented:
Well, `compose_state` could likely switch to storing stream IDs, at least with !is_zephyr_mirror_realm.
